pkg/virtual/framework/forwardingregistry/store.go

```diff
@@ -87,9 +92,115 @@ func DefaultDynamicDelegatedStoreFuncs(
 
 		return delegate.Get(ctx, name, *options, subResources...)
 	}
-	s.CreaterFunc = nil           // not currently supported
-	s.GracefulDeleterFunc = nil   // not currently supported
-	s.CollectionDeleterFunc = nil // not currently supported
+	s.CreaterFunc = func(ctx context.Context, obj runtime.Object, createValidation rest.ValidateObjectFunc, options *metav1.CreateOptions) (runtime.Object, error) {
+		err := rest.BeforeCreate(createStrategy, ctx, obj)
+		if err != nil {
+			return nil, err
+		}
+		if createValidation != nil {
+			if err := createValidation(ctx, obj.DeepCopyObject()); err != nil {
+				return nil, err
+			}
+		}
+		unstructuredObj, ok := obj.(*unstructured.Unstructured)
+		if !ok {
+			return nil, fmt.Errorf("not an Unstructured: %#v", obj)
+		}
+		delegate, err := client(ctx)
+		if err != nil {
+			return nil, err
+		}
+		return delegate.Create(ctx, unstructuredObj, *options, subResources...)
+	}
+	s.GracefulDeleterFunc = func(ctx context.Context, name string, deleteValidation rest.ValidateObjectFunc, options *metav1.DeleteOptions) (runtime.Object, bool, error) {
+		obj, err := s.Get(ctx, name, &metav1.GetOptions{})
+		if err != nil {
+			return nil, false, err
+		}
+		graceful, gracefulPending, err := rest.BeforeDelete(deleteStrategy, ctx, obj, options)
+		if err != nil {
+			return nil, false, err
+		}
+		if deleteValidation != nil {
+			err = deleteValidation(ctx, obj.DeepCopyObject())
+			if err != nil {
+				return nil, false, err
+			}
+		}
+		delegate, err := client(ctx)
+		if err != nil {
+			return nil, false, err
+		}
+		err = delegate.Delete(ctx, name, *options, subResources...)
+		if err != nil {
+			return nil, false, err
+		}
+		return obj, !graceful && !gracefulPending, nil
+	}
+	s.MayReturnFullObjectDeleterFunc = func() bool {
+		return true
+	}
+	s.CollectionDeleterFunc = func(ctx context.Context, deleteValidation rest.ValidateObjectFunc, options *metav1.DeleteOptions, listOptions *metainternalversion.ListOptions) (runtime.Object, error) {
+		list, err := s.List(ctx, listOptions)
+		if err != nil {
+			return nil, err
+		}
+		count := meta.LenList(list)
+		if count == 0 {
+			// Nothing to delete, return now
+			return list, nil
+		}
+
+		workers := goruntime.GOMAXPROCS(0)
+		if workers > count {
+			workers = count
+		}
+
+		objects := make(chan runtime.Object, workers)
+		var deletedObjects []runtime.Object
+
+		group, gCtx := errgroup.WithContext(ctx)
+
+		group.Go(func() error {
+			defer close(objects)
+			return meta.EachListItem(list, func(object runtime.Object) error {
+				objects <- object
+				return nil
+			})
+		})
+
+		for i := 0; i < workers; i++ {
+			group.Go(func() error {
+				for obj := range objects {
+					accessor, err := meta.Accessor(obj)
+					if err != nil {
+						return err
+					}
+					deleted, _, err := s.Delete(gCtx, accessor.GetName(), deleteValidation, options.DeepCopy())
```

Comment:
why deepcopy?

don't we need some uid+rv pre-condition?

Reply:
> why deepcopy?

Individual graceful deleter calls may mutate the options in the delete strategy function calls, so the mutation for one resource of the collection would apply to another one.

> don't we need some uid+rv pre-condition?

The options `UID` and `ResourceVersion` pre-conditions are checked in the `rest.BeforeDelete` function called in the individual graceful deleter function.

Reply:
same comment as above for the normal Delete call: checking in BeforeDelete is not enough.

Reply:
I think we can assume that. Can we somehow hard fail if somebody tries to configure them?

Reply:
and when we need a non-NoOp strategy, we can add another variant that supports it. But the cost is pretty high with the double round-trip.

Reply:
That sounds good. Currently the default REST storage strategy is used in these places:
* APIExport virtual workspace: https://github.com/kcp-dev/kcp/blob/d76f364c549fea913f87c8458a773337c35a9d1d/pkg/virtual/apiexport/builder/forwarding.go#L48-L57
* Syncer virtual workspace: https://github.com/kcp-dev/kcp/blob/d76f364c549fea913f87c8458a773337c35a9d1d/pkg/virtual/syncer/builder/forwarding.go#L48-L57
* Initializing workspace: https://github.com/kcp-dev/kcp/blob/d76f364c549fea913f87c8458a773337c35a9d1d/pkg/virtual/initializingworkspaces/builder/forwarding.go#L45-L54

Generally, it's easy to have it removed and use a default "strategy-less" forwarding store, but can you confirm it's ok to apply it to these 3 places, or should some of them still be configured with a "strategy-aware" forwarding store?

Reply:
Note: we want the strategy for validation, i.e. for write requests. Just not for read.

Reply:
This is still open and IMO a nogo for the PR. It creates a 1:n fanout.

Reply:
Please see here: https://gist.github.com/808ccd44e3ff5d7e1d32c17e616046fe

With that you can compare to `NullAdmission` in these funcs, and then either panic or implement both cases.